File: app/src/main/java/com/boardgamegeek/ui/GamePlaysFragment.java
package com.boardgamegeek.ui;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v4.widget.SwipeRefreshLayout.OnRefreshListener;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.boardgamegeek.R;
import com.boardgamegeek.entities.RefreshableResource;
import com.boardgamegeek.entities.Status;
import com.boardgamegeek.provider.BggContract;
import com.boardgamegeek.ui.model.Game;
import com.boardgamegeek.ui.model.PlaysByGame;
import com.boardgamegeek.ui.viewmodel.GameViewModel;
import com.boardgamegeek.ui.widget.TimestampView;
import com.boardgamegeek.util.PaletteUtils;
import com.boardgamegeek.util.PresentationUtils;

import java.util.List;

import butterknife.BindView;
import butterknife.BindViews;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

public class GamePlaysFragment extends Fragment {
	private static final String KEY_GAME_ID = "GAME_ID";
	private static final String KEY_GAME_NAME = "GAME_NAME";

	private int gameId;
	private String gameName;
	private String imageUrl;
	private String thumbnailUrl;
	private String heroImageUrl;
	private boolean arePlayersCustomSorted;
	@ColorInt private int iconColor = Color.TRANSPARENT;
	private GameViewModel viewModel;

	Unbinder unbinder;
	@BindView(R.id.swipe_refresh) SwipeRefreshLayout swipeRefreshLayout;
	@BindView(R.id.plays_root) View playsRoot;
	@BindView(R.id.plays_label) TextView playsLabel;
	@BindView(R.id.plays_last_play) TextView lastPlayView;
	@BindView(R.id.play_stats_root) View playStatsRoot;
	@BindView(R.id.colors_root) View colorsRoot;
	@BindView(R.id.game_colors_label) TextView colorsLabel;
	@BindView(R.id.sync_timestamp) TimestampView syncTimestampView;
	@BindViews({
		R.id.icon_plays,
		R.id.icon_play_stats,
		R.id.icon_colors
	}) List<ImageView> colorizedIcons;

	public static GamePlaysFragment newInstance(int gameId, String gameName) {
		Bundle args = new Bundle();
		args.putInt(KEY_GAME_ID, gameId);
		args.putString(KEY_GAME_NAME, gameName);
		GamePlaysFragment fragment = new GamePlaysFragment();
		fragment.setArguments(args);
		return fragment;
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		readBundle(getArguments());
	}

	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		return inflater.inflate(R.layout.fragment_game_plays, container, false);
	}

	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		unbinder = ButterKnife.bind(this, view);

		swipeRefreshLayout.setOnRefreshListener(new OnRefreshListener() {
			@Override
			public void onRefresh() {
				viewModel.refresh();
			}
		});
		swipeRefreshLayout.setColorSchemeResources(PresentationUtils.getColorSchemeResources());

		viewModel = ViewModelProviders.of(getActivity()).get(GameViewModel.class);
		viewModel.getGame().observe(this, new Observer<RefreshableResource<Game>>() {
			@Override
			public void onChanged(@Nullable RefreshableResource<Game> game) {
				onGameQueryComplete(game.getData());
			}
		});

		viewModel.getPlays().observe(this, new Observer<RefreshableResource<PlaysByGame>>() {
			@Override
			public void onChanged(@Nullable RefreshableResource<PlaysByGame> playsByGame) {
				if (playsByGame == null) return;
				updateRefreshStatus(playsByGame.getStatus() == Status.REFRESHING);
				onPlaysQueryComplete(playsByGame.getData());
			}
		});

		viewModel.getPlayColors().observe(this, new Observer<List<String>>() {
			@Override
			public void onChanged(@Nullable List<String> strings) {
				int count = strings == null ? 0 : strings.size();
				colorsLabel.setText(PresentationUtils.getQuantityText(getContext(), R.plurals.colors_suffix, count, count));
				colorsRoot.setVisibility(View.VISIBLE);
			}
		});
	}

	private void readBundle(@Nullable Bundle bundle) {
		if (bundle == null) return;
		gameId = bundle.getInt(KEY_GAME_ID, BggContract.INVALID_ID);
		gameName = bundle.getString(KEY_GAME_NAME);
	}

	@Override
	public void onDestroyView() {
		super.onDestroyView();
		if (unbinder != null) unbinder.unbind();
	}

	protected void updateRefreshStatus(final boolean refreshing) {
		if (swipeRefreshLayout != null) {
			swipeRefreshLayout.post(new Runnable() {
				@Override
				public void run() {
					if (swipeRefreshLayout != null) swipeRefreshLayout.setRefreshing(refreshing);
				}
			});
		}
	}

	private void onGameQueryComplete(Game game) {
		gameName = game.getName();
		imageUrl = game.getImageUrl();
		thumbnailUrl = game.getThumbnailUrl();
		heroImageUrl = game.getHeroImageUrl();
		arePlayersCustomSorted = game.getCustomPlayerSort();
		syncTimestampView.setTimestamp(game.getUpdatedPlays());
		iconColor = game.getIconColor();
		colorize();
	}

	private void onPlaysQueryComplete(PlaysByGame plays) {
		if (plays != null) {
			playsRoot.setVisibility(VISIBLE);

			String description = PresentationUtils.describePlayCount(getActivity(), plays.getPlayCount());
			if (!TextUtils.isEmpty(description)) {
				description = " (" + description + ")";
			}
			playsLabel.setText(PresentationUtils.getQuantityText(getActivity(), R.plurals.plays_prefix, plays.getPlayCount(), plays.getPlayCount(), description));

			if (plays.getMaxDate() > 0) {
				lastPlayView.setText(PresentationUtils.getText(getActivity(), R.string.last_played_prefix, PresentationUtils.describePastDaySpan(plays.getMaxDate())));
				lastPlayView.setVisibility(VISIBLE);
			} else {
				lastPlayView.setVisibility(GONE);
			}

			playStatsRoot.setVisibility(plays.getPlayCount() == 0 ? GONE : VISIBLE);
		}
	}

	private void colorize() {
		if (isAdded() &&
			iconColor != Color.TRANSPARENT &&
			colorizedIcons != null) {
			ButterKnife.apply(colorizedIcons, PaletteUtils.getRgbIconSetter(), iconColor);
		}
	}

	@OnClick(R.id.plays_root)
	public void onPlaysClick() {
		GamePlaysActivity.start(getContext(), gameId, gameName, imageUrl, thumbnailUrl, heroImageUrl, arePlayersCustomSorted, iconColor);
	}

	@OnClick(R.id.play_stats_root)
	public void onPlayStatsClick() {
		GamePlayStatsActivity.start(getContext(), gameId, gameName, iconColor);
	}

	@OnClick(R.id.colors_root)
	public void onColorsClick() {
		GameColorsActivity.start(getContext(), gameId, gameName, iconColor);
	}
}

File: app/src/main/java/com/boardgamegeek/ui/GamePlaysFragment.kt
package com.boardgamegeek.ui

import android.annotation.SuppressLint
import android.os.Bundle
import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.pluralStringResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import com.boardgamegeek.R
import com.boardgamegeek.databinding.FragmentGamePlaysBinding
import com.boardgamegeek.extensions.addAlphaToColor
import com.boardgamegeek.extensions.asPlayCount
import com.boardgamegeek.extensions.isKnownColor
import com.boardgamegeek.extensions.isToday
import com.boardgamegeek.extensions.toFormattedString
import com.boardgamegeek.model.Play
import com.boardgamegeek.provider.BggContract
import com.boardgamegeek.ui.compose.*
import com.boardgamegeek.ui.viewmodel.GameViewModel
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.delay
import kotlin.time.Duration.Companion.seconds

@AndroidEntryPoint
class GamePlaysFragment : Fragment() {
    private var _binding: FragmentGamePlaysBinding? = null
    private val binding get() = _binding!!

    @Suppress("RedundantNullableReturnType")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentGamePlaysBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

//        binding.swipeRefresh.setOnRefreshListener { viewModel.refreshPlays() }

        binding.composeView.setContent {
            val viewModel by activityViewModels<GameViewModel>()
            val game by viewModel.game.observeAsState()
            val plays by viewModel.plays.observeAsState(emptyList())
            val colors by viewModel.playColors.observeAsState(emptyList())
            val isRefreshing by viewModel.playsAreRefreshing.observeAsState(false)

            binding.swipeRefresh.isRefreshing = isRefreshing
            if (isRefreshing) {
                BggLoadingIndicatorBox()
            }

            game?.let { game ->
                if (game.id == BggContract.INVALID_ID) {
                    ErrorContent(
                        stringResource(R.string.invalid_id),
                        painterResource(R.drawable.game_24px)
                    )
                } else {
                    Column {
                        val iconColor = Color(game.iconColor.addAlphaToColor())
                        Spacer(Modifier.height(8.dp))
                        TotalPlaysRow(plays) {
                            GamePlaysActivity.start(
                                requireContext(),
                                game.id,
                                game.name,
                                game.heroImageUrl,
                                game.thumbnailUrl,
                                game.customPlayerSort,
                                game.iconColor,
                            )
                        }
                        val inProgressPlays = plays.filter { play -> play.dirtyTimestamp > 0L }
                        if (inProgressPlays.isNotEmpty()) {
                            InProgressPlaysRow(inProgressPlays, iconColor) { play ->
                                PlayActivity.start(requireContext(), play.internalId)
                            }
                        }
                        plays.filter { play -> play.dirtyTimestamp == 0L }.maxByOrNull { play -> play.dateInMillis }?.let { lastPlay ->
                            LastPlayRow(lastPlay, iconColor) {
                                PlayActivity.start(requireContext(), lastPlay.internalId)
                            }
                        }
                        if (plays.isNotEmpty()) {
                            StatsRow(iconColor) {
                                GamePlayStatsActivity.start(requireContext(), game.id, game.name, game.iconColor)
                            }
                        }
                        ColorsRow(colors, iconColor) {
                            GameColorsActivity.start(requireContext(), game.id, game.name, game.iconColor)
                        }
                        GameFooter(
                            game.updatedPlays,
                            game.id,
                            Modifier
                                .fillMaxWidth()
                                .padding(horizontal = dimensionResource(R.dimen.material_margin_horizontal))
                        )
                    }
                }
            } ?: EmptyContent(
                stringResource(R.string.empty_plays_game),
                painterResource(R.drawable.game_24px),
            )
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}

@Composable
fun TotalPlaysRow(
    plays: List<Play>,
    onClick: () -> Unit = { },
) {
    val playCount = plays.sumOf { it.quantity }
    val (count, description, color) = playCount.asPlayCount(LocalContext.current)
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier
            .fillMaxWidth()
            .heightIn(dimensionResource(R.dimen.game_row_height))
            .clickable(onClick = onClick)
            .padding(horizontal = dimensionResource(R.dimen.material_margin_horizontal))
    ) {
        ColorBox(
            color,
            size = ColorBoxDefaults.sizeSmall,
            modifier = Modifier.padding(start = 0.dp, end = 24.dp),
        ) {
            Text(count.toFormattedString(), color = it, modifier = Modifier.align(Alignment.Center))
        }
        Column(
            modifier = Modifier
                .padding(vertical = 4.dp)
                .heightIn(dimensionResource(R.dimen.game_row_height)),
            verticalArrangement = Arrangement.Center,
        ) {
            PrimaryRowText(pluralStringResource(R.plurals.play_title_suffix, playCount, playCount))
            if (description.isNotBlank()) SecondaryRowText(description)
        }
    }
}

@SuppressLint("LocalContextGetResourceValueCall")
@Composable
private fun InProgressPlaysRow(
    plays: List<Play>,
    iconColor: Color = Color.Transparent,
    onClick: (Play) -> Unit = { },
) {
    GameRow(
        R.drawable.ic_outline_timer_24,
        R.string.title_in_progress,
        iconColor = iconColor,
    ) {
        Column {
            plays.take(5).forEach { play ->
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.Start,
                    modifier = Modifier
                        .fillMaxWidth()
                        .heightIn(min = 60.dp)
                        .clickable(onClick = { onClick(play) })
                ) {
                    Column(
                        verticalArrangement = Arrangement.Center,
                        horizontalAlignment = Alignment.Start,
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        val context = LocalContext.current
                        var headline by remember { mutableStateOf("") }
                        LaunchedEffect(Unit) {
                            while (true) {
                                headline = if (play.startTime > 0) {
                                    context.getString(
                                        R.string.playing_for_prefix,
                                        DateUtils.formatElapsedTime((System.currentTimeMillis() - play.startTime) / 1000)
                                    )
                                } else if (play.dateInMillis.isToday()) {
                                    context.getString(R.string.playing_prefix, play.dateForDisplay(context))
                                } else {
                                    context.getString(R.string.playing_since_prefix, play.dateForDisplay(context))
                                }
                                delay(30.seconds)
                            }
                        }
                        ListItemPrimaryText(headline)
                        ListItemSecondaryText(play.describe(LocalContext.current, false))
                    }
                }
            }
        }
    }
}

@Composable
private fun LastPlayRow(
    play: Play,
    iconColor: Color = Color.Transparent,
    onClick: () -> Unit = { },
) {
    GameRow(
        R.drawable.plays_24px,
        R.string.last_played,
        Modifier.clickable(onClick = onClick),
        iconColor = iconColor,
    ) {
        SecondaryRowText(stringResource(R.string.last_played))
        PlayListItem(
            play,
            showGameName = false,
            padding = PaddingValues(0.dp),
            minimumHeight = ListItemDefaults.oneLineHeight,
        )
    }
}

@Composable
private fun StatsRow(
    iconColor: Color = Color.Transparent,
    onClick: () -> Unit = { },
) {
    GameRow(
        R.drawable.play_stats_24px,
        R.string.title_play_stats,
        Modifier
            .semantics(mergeDescendants = true) {
                role = Role.Button
            }
            .clickable(onClick = onClick),
        iconColor = iconColor,
    ) {
        PrimaryRowText(stringResource(R.string.title_play_stats))
    }
}

@Composable
private fun ColorsRow(
    colors: List<String>,
    iconColor: Color = Color.Transparent,
    onClick: () -> Unit = { },
) {
    GameRow(
        R.drawable.color_24px,
        R.string.colors,
        Modifier.clickable(onClick = onClick),
        iconColor = iconColor,
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
        ) {
            PrimaryRowText(
                pluralStringResource(R.plurals.colors_suffix, colors.size, colors.size),
                modifier = Modifier.weight(1f),
            )
            if (colors.isNotEmpty() && colors.all { it.isKnownColor() }) {
                colors.forEach {
                    ColorBox(it, size = ColorBoxDefaults.sizeSmall)
                }
            }
        }
    }
}

@Preview(backgroundColor = 0xFFFFFFFF, showBackground = true)
@Composable
private fun GamePlaysPreview() {
    Column {
        val iconColor = Color.Blue
        TotalPlaysRow(emptyList())
        TotalPlaysRow(
            listOf(
                Play(
                    dateInMillis = System.currentTimeMillis(),
                    gameId = 13,
                    gameName = "Terra Mystica",
                ),
            ),
        )
        InProgressPlaysRow(
            listOf(
                Play(
                    internalId = BggContract.INVALID_ID.toLong(),
                    playId = BggContract.INVALID_ID,
                    dateInMillis = System.currentTimeMillis(),
                    gameId = 13,
                    gameName = "Terra Mystica",
                    quantity = 1,
                    location = "Library",
                    startTime = System.currentTimeMillis() - 12345L
                ),
                Play(
                    internalId = BggContract.INVALID_ID.toLong(),
                    playId = BggContract.INVALID_ID,
                    dateInMillis = System.currentTimeMillis(),
                    gameId = 13,
                    gameName = "7th Continent",
                    quantity = 1,
                    length = 127,
                ),
                Play(
                    internalId = BggContract.INVALID_ID.toLong(),
                    playId = BggContract.INVALID_ID,
                    dateInMillis = System.currentTimeMillis() - 99999999L,
                    gameId = 13,
                    gameName = "HUANG",
                    quantity = 1,
                    location = "House",
                    length = 98,
                ),
            ),
            iconColor
        )
        LastPlayRow(
            Play(
                internalId = BggContract.INVALID_ID.toLong(),
                playId = BggContract.INVALID_ID,
                dateInMillis = System.currentTimeMillis(),
                gameId = 13,
                gameName = "Catan",
                quantity = 1,
                length = 46,
                location = "House",
            ),
            iconColor,
        )
        StatsRow(iconColor)
        ColorsRow(listOf("Blue", "Yellow"), iconColor)
    }
}